### PR TITLE
Complete base models and serializers

### PR DESCRIPTION
**Pull request by imsickofmaps**

### CODE REVIEW

File: contentstore/models.py [hunk 1]
```diff
@@ -3,6 +3,12 @@
 
 
 class Schedule(models.Model):
+
+    """
+    Schdules (sometimes referred to as Protocols) are the method used to
```
Comment (smn) [inline contentstore/models.py]: typo

**Comment by imsickofmaps** [comment]
@smn I've finished with this one now, not going to keep adding stuff it else it'll get too large. Please review and I'll start on the verified fake.

File: contentstore/models.py [hunk 2]
```diff
@@ -1,8 +1,17 @@
 from django.db import models
+# from django.core.exceptions import ValidationError
```
Comment (smn) [inline contentstore/models.py]: Remove?

File: contentstore/models.py [hunk 3]
```diff
@@ -27,3 +36,79 @@ def __unicode__(self):
...
+                                 null=True,
+                                 blank=True)
+    default_schedule = models.ForeignKey(Schedule,
+                                         related_name='message_sets',
+                                         null=False)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    def __unicode__(self):
+        return "%s" % self.short_name
```
Comment (smn) [inline contentstore/models.py]: You're returning a byte string here, not unicode.

File: contentstore/models.py [hunk 4]
```diff
@@ -27,3 +36,79 @@ def __unicode__(self):
...
+
+
+def generate_new_filename(instance, filename):
+    ext = filename.split('.')[-1]  # get file extension
```
Comment (smn) [inline contentstore/models.py]:
``` python
>>> import os.path
>>> os.path.splitext('foo.bar')
('baz/foo', '.bar')
```

File: contentstore/models.py [hunk 5]
```diff
@@ -27,3 +36,79 @@ def __unicode__(self):
...
+        settings file.
+    """
+
+    content = models.FileField(upload_to=generate_new_filename,
+                               max_length=100)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    def __unicode__(self):
+        return "%s" % (self.content.path.split('/')[-1])
```
Comment (smn) [inline contentstore/models.py]: Not returning unicode, also:

``` python
>>> os.path.basename('foo/bar.baz')
'bar.baz'
```

File: contentstore/models.py [hunk 6]
```diff
@@ -27,3 +36,79 @@ def __unicode__(self):
...
+
+
+class Message(models.Model):
+
+    """
+        Messages that a recipient can be sent
+    """
+    messageset = models.ForeignKey(MessageSet,
+                                   related_name='messages',
```
Comment (smn) [inline contentstore/models.py:88]: Would `messages` as a related name not be confusing here? Considering we're pointing at the name of a set, not the actual messages?

File: contentstore/models.py [hunk 7]
```diff
@@ -27,3 +36,79 @@ def __unicode__(self):
...
+                                       null=True)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        ordering = ['sequence_number']
+
+    def clean(self):
+        # Don't allow messages to have neither a text or binary content
+        if self.text_content is None and self.binary_content is None:
```
Comment (smn) [inline contentstore/models.py]: bikeshed on not using `any()` here.

File: contentstore/models.py [hunk 8]
```diff
@@ -27,3 +36,79 @@ def __unicode__(self):
...
+            raise ValidationError(
+                _('Messages must have text or file attached'))
+
+    def save(self, *args, **kwargs):
+        self.clean()
+        super(Message, self).save(*args, **kwargs)
```
Comment (smn) [inline contentstore/models.py:110]: Should we return the value of the `super()` call here?

File: contentstore/models.py [hunk 9]
```diff
@@ -27,3 +36,79 @@ def __unicode__(self):
...
+
+    def __unicode__(self):
+        return _("Message %s in %s from %s") % (
+            self.sequence_number, self.lang, self.messageset.short_name)
```
Comment (smn) [inline contentstore/models.py:114]: Not sure if `_` forces unicode here or not.

File: contentstore/tests.py [hunk 10]
```diff
@@ -29,7 +29,45 @@ def setUp(self):
...
+            messageset=messageset, sequence_number=sequence_number,
+            lang=lang, text_content=text_content,
+            binary_content=binary_content)
+        return message
+
+    def make_binary_content(self):
+        # models.generate_new_filename = lambda *a: "20151201010101012345.png"
+        simple_png = BytesIO(
+            b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x01\x00\x00\x00\x01\x08\x06\x00\x00\x00\x1f\x15\xc4\x89\x00\x00\x00\rIDATx\x9cc````\x00\x00\x00\x05\x00\x01\xa5\xf6E@\x00\x00\x00\x00IEND\xaeB`\x82')   # flake8: noqa
+        simple_png.name = 'test.png'
```
Comment (smn) [inline contentstore/tests.py]: I'd just read a sample `.png` from disk and use `pkg_resources.resource_stream('pkg_name', 'filename')` to get it from this package's location on disk.

**Comment by smn** [comment]
Minor stuff, just curious about the tests posting to `/schedule/`. Scheduling is in a different thing now. Was this perhaps copied over from momconnect?

File: contentstore/models.py (hunk 8 above)
Comment (imsickofmaps) [inline contentstore/models.py:110]: It's "called" not returned. See: https://docs.djangoproject.com/en/1.8/topics/db/models/#overriding-model-methods

File: contentstore/models.py (hunk 9 above)
Comment (imsickofmaps) [inline contentstore/models.py:114]: It does. See note here: https://docs.djangoproject.com/en/1.8/topics/i18n/translation/#standard-translation

File: contentstore/models.py (hunk 6 above)
Comment (imsickofmaps) [inline contentstore/models.py:88]: `related_name` is used from the relation back to this one. See: https://docs.djangoproject.com/en/1.8/ref/models/fields/#django.db.models.ForeignKey.related_name 
We do this so the REST serializer is consistent with its returned relations array.

**Comment by imsickofmaps** [comment]
Regarding `/schedule/` we can discuss in detail but my assumption from reading the spec was that the content store still needed to be able to CRUD "protocols" which are best generically defined as schedules.

**Comment by smn** [comment]
:+1: on the changes. We've had a separate scheduler app developed by @praekelt/lyndsay specifically for the scheduling (to pull that out of the content store) so let's chat about that to make sure there's no disconnect there.

**Pull request merged**